src/list.cpp

```diff
@@ -286,9 +258,79 @@ void list_probes(const BPFtrace &bpftrace, const std::string &search_input)
 
     std::cout << probe << std::endl;
   }
+}
+
+void list_probes(const BPFtrace& bpftrace, const std::string& search_input)
+{
+  std::string search = search_input;
+  std::string probe_name;
+  bool has_wildcard_in_probe = false;
+  std::regex re;
+
+  std::smatch probe_match;
+  std::regex probe_regex(":.*");
+  std::regex_search(search, probe_match, probe_regex);
+
+  // replace alias name with full name
+  if (probe_match.size())
+  {
+    auto pos = probe_match.position(0);
+    probe_name = probetypeName(search.substr(0, probe_match.position(0)));
+    search = probe_name + search.substr(pos, search.length());
+    for (char c : probe_name)
+      has_wildcard_in_probe |= c == '*' || c == '?';
+  }
+
+  std::string s = "^";
+  for (char c : search)
+  {
+    if (c == '*')
+      s += ".*";
+    else if (c == '?')
+      s += '.';
+    else
+      s += c;
+  }
+  s += '$';
+  try
+  {
+    re = std::regex(s,
+                    std::regex::icase | std::regex::grep | std::regex::nosubs |
+                        std::regex::optimize);
+  }
+  catch (std::regex_error& e)
+  {
+    std::cerr << "ERROR: invalid character in search expression." << std::endl;
+    return;
+  }
+
+  // software
+  if (has_wildcard_in_probe || probe_name.empty() || probe_name == "software")
```

Comment:
Maybe pull out `has_wildcard_in_probe || probe_name.empty()` into another variable to repeat less code